### PR TITLE
Push on several ipfs nodes

### PR DESCRIPTION
**Pull request by yann300**
This allow to push on several ipfs gateway

### CODE REVIEW

File: src/lib/publishOnIpfs.js [hunk 1]
```diff
@@ -99,3 +102,28 @@ async function ipfsVerifiedPublish (content, expectedHash, cb) {
     cb(error)
   }
 }
+
+async function tryPushes (content) {
+  return new Promise((resolve, reject) => {
```
Comment (GrandSchtroumpf) [inline src/lib/publishOnIpfs.js]: As you're using `async` you don't need to create a promise. 
```javascript
await function goodFunctionName(content) {
  let error;  // Do that only if you want to throw the last error after every node have failed
  for (const node of ipfsNodes) {
    try {
      const result = await node.add(content)
      return result  // will stop when first works
    } catch (err) {
      error = err
    }
  }
  throw new Error(error);
}
```

File: src/lib/publishOnIpfs.js [hunk 2]
```diff
@@ -99,3 +102,28 @@ async function ipfsVerifiedPublish (content, expectedHash, cb) {
...
+    let failedCount = 0
+    let alreadyReturned = false
+    const failed = (error) => {
+      console.log(error)
+      if (alreadyReturned) return
+      failedCount++
+      if (failedCount >= ipfsNodes.length) {
```
[the remark this answers is not in the document]
Comment (yann300) [reply src/lib/publishOnIpfs.js]: There is a `forEach`, but we don't `await` for the response

File: src/lib/publishOnIpfs.js (hunk 1 above)
Comment (yann300) [reply src/lib/publishOnIpfs.js]: in the code above, we shouldn't `await`

Comment (yann300) [reply src/lib/publishOnIpfs.js]: removed the `async` keyword, just keeping the `Promise`

Comment (GrandSchtroumpf) [reply src/lib/publishOnIpfs.js]: Why we shouldn't await ?
From what I understand you're trading **readibility** for a small leap of performance (your array is very small).

File: src/lib/publishOnIpfs.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (GrandSchtroumpf) [reply src/lib/publishOnIpfs.js]: But you trigger the `failed` function maximum one time per item in the array, so there is no way the `failedCount` is above the length.

File: src/lib/publishOnIpfs.js (hunk 1 above)
Comment (yann300) [reply src/lib/publishOnIpfs.js]: We are sending requests in parallel and waiting for response:
if one response is succeed, we are good. If we got X errors (X being the amount of IPFS nodes), we fail.

Comment (GrandSchtroumpf) [reply src/lib/publishOnIpfs.js]: I understand that, but the current implementation is very difficult to understand.
If you really want to use concurrency, you can do Promise.all : 
```javascript
const requests = ipfsNodes.map(node => node.add(content).catch(e => e));
const results = await Promise.all(requests);  // Wait for all requests to arrive
const success = results.find(result => !(result instanceof Error));  // Find the first working
success
  ? return success
  : throw new Error(results[0]);
```

Comment (yann300) [reply src/lib/publishOnIpfs.js]: That would be sweet to use `Promise.all`, but that will not work by using it the classic way: 

`Promise.all is rejected if any of the elements are rejected. For example, if you pass in four promises that resolve after a timeout and one promise that rejects immediately, then Promise.all will reject immediately.`
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all#Promise.all_fail-fast_behaviour

checking if we could tweak it

Comment (GrandSchtroumpf) [reply src/lib/publishOnIpfs.js]: That's why I do `.catch(e => e)`. It feel the result with errors

Comment (yann300) [reply src/lib/publishOnIpfs.js]: Ok thanks will have a look shortly. 
Just to be sure: does this handle the fact that even if one request fail, `Promise.all` still wait for the other request to finish before resolving

Comment (GrandSchtroumpf) [reply src/lib/publishOnIpfs.js]: Yes it should. See second answers of this stakeoverflow question: https://stackoverflow.com/questions/30362733/handling-errors-in-promise-all

Comment (GrandSchtroumpf) [reply src/lib/publishOnIpfs.js]: Again, I think you are readibility for a small amount of performance. I had to try to rewrite the code to understand what it was doing. And I think this is bad practice, espacially for a team like remix.

Comment (yann300) [reply src/lib/publishOnIpfs.js]: Well not really, the infura gateway can reply in 10s / 20s sometimes (even more sometimes), whether our node reply in a few ms.
That the function misses some comment is another thing.

**Pull request merged**